### PR TITLE
Fixes #583 Generating a unique cookie name for an application.

### LINKED ISSUE #583: FileResult

### PR DESCRIPTION
**Pull request by harshgMSFT**
@GrabYourPitchforks .. please let me know if this does not sound right.

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgery.cs [hunk 1]
```diff
@@ -24,9 +28,13 @@ public sealed class AntiForgery
                            [NotNull] IDataProtectionProvider dataProtectionProvider,
                            [NotNull] IAntiForgeryAdditionalDataProvider additionalDataProvider,
                            [NotNull] IOptions<MvcOptions> mvcOptions,
-                           [NotNull] IHtmlEncoder htmlEncoder)
+                           [NotNull] IHtmlEncoder htmlEncoder,
+                           IOptions<DataProtectionOptions> dataProtectionOptions)
         {
             var config = mvcOptions.Options.AntiForgeryOptions;
+            var applicationId = dataProtectionOptions.Options.ApplicationDiscriminator ?? string.Empty;
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgery.cs]: Should we use something stronger than `string.Empty` here?

File: src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgery.cs [hunk 2]
```diff
@@ -109,5 +117,15 @@ public void SetCookieTokenAndHeader([NotNull] HttpContext context)
...
             _worker.SetCookieTokenAndHeader(context);
         }
+
+        private string ComputeCookieName(string applicationId)
+        {
+            using (var sha256 = SHA256.Create())
+            {
+                var hash = sha256.ComputeHash(Encoding.UTF8.GetBytes(applicationId));
+                var subHash = hash.Take(4).ToArray();
+                return Convert.ToBase64String(subHash);
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgery.cs]: Just 4 bytes?

File: src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgery.cs (hunk 1 above)
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgery.cs]: nope, the cookie name is not the secret, its the payload that is.

File: src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgery.cs (hunk 2 above)
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.Core/AntiForgery/AntiForgery.cs]: To keep the name short and avoid it being unnecessarily long.

**Comment by harshgMSFT** [comment]
Updated

File: test/Microsoft.AspNet.Mvc.FunctionalTests/AntiForgeryTestHelper.cs [hunk 3]
```diff
@@ -51,13 +51,25 @@ public static string RetrieveAntiForgeryToken(string htmlContent, string actionU
...
+                Key = cookieKey,
+                Value = cookieData
+            };
+        }
+
+        public class CookieMetadata
+        {
+            public string Key { get; set; }
+
+            public string Value { get; set; }
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.FunctionalTests/AntiForgeryTestHelper.cs]: Just re-use `KeyValuePair` :laughing:

**Comment by rynowak** [comment]
:shipit:

**Pull request closed without merging**